torch_geometric/nn/models/rbcd_attack.py

```diff
@@ -0,0 +1,760 @@
+        metric: Optional[Union[str, LOSS_TYPE]] = None,
+        lr: float = 1_000,
+        is_undirected_graph: bool = True,
+        log: bool = True,
+        **kwargs,
+    ):
+        super().__init__()
+
+        self.model = model
+        self.block_size = block_size
+        self.epochs = epochs
+
+        if isinstance(loss, str):
+            if loss == 'masked':
+                self.loss = self._masked_cross_entropy
+            elif loss == 'margin':
+                self.loss = partial(self._margin_loss, reduce='mean')
+            elif loss == 'prob_margin':
+                self.loss = self._probability_margin_loss
+            elif loss == 'tanh_margin':
+                self.loss = self._tanh_margin_loss
+            else:
+                raise ValueError(f'Unknown loss `{loss}`')
+        else:
+            self.loss = loss
+
+        self.is_undirected_graph = is_undirected_graph
+        self.log = log
+
+        if metric is None:
+            self.metric = self.loss
+        else:
+            self.metric = metric
+
+        self.epochs_resampling = epochs_resampling
+        self.lr = lr
+
+        self.coeffs.update(kwargs)
+
+    def attack(self, x: Tensor, edge_index: Tensor, labels: Tensor,
+               budget: int, idx_attack: Optional[Tensor] = None,
+               **kwargs) -> Tuple[Tensor, Tensor]:
+        """Attack the predictions for the provided model and graph.
+
+        A subset of predictions may be specified with :attr:`idx_attack`. The
+        attack is allowed to flip (i.e. add or delete) :attr:`budget` edges and
+        will return the strongest perturbation it can find. It returns both the
+        resulting perturbed :attr:`edge_index` as well as the perturbations.
+
+        Args:
+            x (Tensor): The node feature matrix. We assume `x` to be located
+                on target device.
+            edge_index (LongTensor): The edge indices.
+            labels (Tensor): The labels.
+            budget (int): The number of allowed perturbations (i.e.
+                number of edges that are flipped at most).
+            idx_attack (Tensor, optional): Filter for predictions/labels.
+                Shape and type must match that it can index :attr:`labels`
+                and the model's predictions.
+            **kwargs (optional): Additional arguments passed to the GNN module.
+
+        :rtype: (:class:`Tensor`, :class:`Tensor`)
+        """
+        assert self.block_size > budget, (
+            f'The search space size ({self.block_size}) must be '
+            f'greater than the number of permutations ({budget})')
+
+        self.model.eval()
+
+        self.device = x.device
+        assert kwargs.get('edge_weight', None) is None, \
+            '`edge_weight` is not supported'
+        edge_weight = torch.ones(edge_index.size(1), device=self.device)
+        self.edge_index = edge_index.cpu().clone()
+        self.edge_weight = edge_weight.cpu().clone()
```

Comment:
Any reason these need to be on CPU? Why do we create `edge_weight` on the GPU and convert it to CPU afterwards?

Reply:
See last week's high-level comment